prog/dftb+/lib_timedep/linrespcommon.F90

```diff
@@ -535,66 +606,285 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
+      !$OMP PARALLEL DO DEFAULT(SHARED) PRIVATE(ia,ii,jj,ss,qij) &
       !$OMP& SCHEDULE(RUNTIME)
       do ia = 1,nmat
 
         qij(:) = transChrg%qTransIA(ia, iAtomStart, stimc, grndEigVecs, getia, win)
 
-        updwn = (win(ia) <= nmatup)
-        if (updwn) then
-          fact = 1.0_dp
-        else
-          fact =-1.0_dp
-        end if
-        vout(ia) = vout(ia) + fact * wnij(ia) * dot_product(qij, otmp)
+        ss = getIA(win(ia), 3)
+
+        vout(ia) = vout(ia) + 2.0_dp * sqrOccIA(ia) * spinFactor(ss) * dot_product(qij, otmp)
 
       end do
       !$OMP  END PARALLEL DO
+      
 
     end if
 
     if (allocated(onsMEs)) then
-      call onsiteEner(spin, sym, wnij, win, nmatup, iAtomStart, getia, species0, stimc,&
-          & grndEigVecs, onsMEs, orb, vin, vout_ons)
+      call onsiteEner(spin, sym, wij, sqrOccIA, win, nmatup, iAtomStart, getia, species0, stimc,&
+          & grndEigVecs, onsMEs, orb, vTmp, vout_ons)
       vout(:) = vout + vout_ons
     end if
 
-    vout(:) = vout(:) + ( wij**2 ) * vin
+    vout(:) = vout(:) + wij(:) * vTmp(:)
+
+    if(.not. tAplusB) then
+       vout(:) = vout(:) * sqrt(wij(:))
+    endif
+
+  end subroutine actionAplusB 
+
+  !> Multiplies the excitation supermatrix with a supervector.
+  !> (A-B)_ias,jbt * v_jbt is computed (and similar for singlet/triplet)
+  !> (see definitions in Marc Casida, in Recent Advances in Density Functional Methods,
+  !>  World Scientific, 1995, Part I, p. 155.)
+  !> See also Dominguez JCTC 9 4901 (2013), Kranz JCTC 13 1737 (2017) for DFTB specifics.
+  subroutine actionAminusB(tSpin, win, nMatUp, occNr, getIA, wIA, vIn, vOut)
+
+    !> logical spin polarization
+    logical, intent(in) :: tSpin
+
+    !> sorting index of the excitation energies.
+    integer, intent(in) :: win(:)
+
+    !> number of same-spin transitions
+    integer, intent(in) :: nMatUp
+
+    !> occupation numbers
+    real(dp), intent(in) :: occNr(:,:)
+
+    !> index array for excitations
+    integer, intent(in) :: getIA(:,:)
+
+    !> excitation energies (wIA = epsilon_a - epsilon_i)
+    real(dp), intent(in) :: wIA(:)
+
+    !> vector to multiply with size(nmat)
+    real(dp), intent(in) :: vIn(:)
+
+    !> vector containing the result on exit size(nmat)
+    real(dp), intent(out) :: vOut(:)
 
-  end subroutine omegatvec
+    integer :: nMat
 
+    nMat = size(vin) 
+    vOut(:) = 0.0_dp
+
+    ! orb. energy difference diagonal contribution
+    vOut(:) = vOut(:) + wIA(:) * vIn(:) 
+
+  end subroutine actionAminusB
+
+  !> Generates initial matrices M+ and M- for the RPA algorithm by Stratmann
+  !> (JCP 109 8218 (1998).
+  !> M+/- = (A+/-B)_ias,jbt (spin polarized) (A+/-B)^{S/T}_ia,jb (closed shell) 
+  !> Here ias,jbt <= initDim
+  !> Also computed is v+/- = (A+/-B)_ias,jbt with ias <= nMat, jbt <= initDim
+  !> Note: Routine not set up to handle onsite corrections.
+  !> Note: Not yet OpenMP parallelized
```

Comment:
As it calls subroutines with OMP parallelisation inside (e.g. `hemv`), probably this routine should not be OpenMP parallelized in its current form anyway...

Reply:
ok